adding remaining heuristics for review

magnoxemo opened the pull request:
closes #5

magnoxemo committed:
adding remaining heuristics for review

gonuke commented on doc/content/source/userobjects/ValueRangeHeuristicUserObject.md:
```diff
@@ -0,0 +1,27 @@
+# ValueRangeHeuristicUserObejct
```
Isn't this just a special case of the boolean combination of two threshold heuristics?  Do we need a separate case or just advise users to use a pair of threshold heuristics?

magnoxemo replied:
Yes combination of two threshold heuristics would do the same thing. But I think value range is more user friendly.

gonuke commented on doc/content/source/userobjects/ValueFractionHeuristicUserObject.md:
```diff
@@ -0,0 +1,29 @@
+# ValueFractionHeuristicUserObejct
```
This seems like a special case of threshold heuristic, or boolean pair of them, depending on how the users wants to use it.

The threshold in this case is defined relative to the min/max, but once the threshold is defined, it's the same.

gonuke replied:
There could be an argument for making a special case, because it's easier for users, but I don't think it more computationally efficient

magnoxemo replied:
The upper and lower fraction changes dynamically in every adaptivity step. This could be more beneficial for filtering out maximum relative error region which ( relative error scale ) changes in every adaptivity step.

gonuke replied:
The threshold could be recalculated at each adaptivity step too, so that's not a limitation

gonuke commented on include/userobjects/ValueDifferenceHeuristicUserObject.h:
```diff
@@ -0,0 +1,19 @@
+protected:
+  /// method for evaluting if element should be clusterd or not
+  virtual bool evaluate(libMesh::Elem * base_element,
+                                libMesh::Elem * neighbor_elem)const override;
```
Does Cardinal use clang-format?  This looks poorly formatted

gonuke commented on src/userobjects/ValueDifferenceHeuristicUserObject.C:
```diff
@@ -0,0 +1,28 @@
+                                                     libMesh::Elem * neighbor_element) const{
+  const Real base_score = getMetricData(base_element);
+  const Real neighbor_score =  getMetricData(neighbor_element);
+  return std::abs((base_score-neighbor_score)/base_score) < _tolerance;
```
I think standard formatting would look like this:
```suggestion
  return std::abs((base_score - neighbor_score) / base_score) < _tolerance;
```
Not sure if Cardinal imposes any style guide on things like this

gonuke reviewed (changes requested):
This looks mostly OK, but some questions about how many specialized versions of heuristics to develop.  I'll take the question to slack

magnoxemo committed:
clang format

gonuke commented on doc/content/source/userobjects/ValueFractionHeuristicUserObject.md:
```diff
@@ -0,0 +1,29 @@
+elements are both either more/less than a upper/lower cut off of the metric variable.
+The upper and lower cut off is calculated is calculated as
+
+upper_cut_off $= (1 -$ upper_fraction $)$ $ \times S_{max}$
+
+lower_cut_off $= $ lower_fraction $ \times (S_{max} - S_{min}) + S_{min}$
```
These are not self-consistent.... why not?
```suggestion
upper_cut_off $= (1 -$ upper_fraction $)$ $ \times (S_{max} - S_{min}) + S_{min}$ $

lower_cut_off $= $ lower_fraction $ \times (S_{max} - S_{min}) + S_{min}$
```

gonuke commented on doc/content/source/userobjects/ValueRangeHeuristicUserObject.md:
```diff
@@ -0,0 +1,27 @@
+$$S_{lower} = (1-tol) \times Value $$
+$$S_{upper} = (1+tol) \times Value $$
+
+where $Value$ and $tol$ are user provided inputs.
```
These are not the names/capitalization used in the code

gonuke commented on src/userobjects/ValueFractionHeuristicUserObject.C:
```diff
@@ -0,0 +1,53 @@
+      _min = score;
+    }
+  }
+  _upper_cut_off = (1 - _upper_fraction) * _max;
+  _lower_cut_off = _lower_fraction * (_max - _min) + _min;
```
Why don't we define these consistently?
```suggestion
  _upper_cut_off = (1 - _upper_fraction) * (_max - _min) + _min;
  _lower_cut_off = _lower_fraction * (_max - _min) + _min;
```

gonuke commented on src/userobjects/ValueFractionHeuristicUserObject.C:
```diff
@@ -0,0 +1,53 @@
+ValueFractionHeuristicUserObject::extremesFinder()
+{
+  double score;
+  for (auto & elem : _mesh.active_element_ptr_range())
```
I'm sure I've asked this before, but does LibMesh not define a mechanism for evaluating the maximum value of a variable across a mesh?

magnoxemo replied:
I don't think so. I have seen they iterate though the solution vector to determine max and min. [here](https://github.com/idaholab/moose/blob/9241cbb0ee63b74d56f7789de1168d9ab9353f10/framework/src/markers/ErrorFractionMarker.C#L53 )

I will ask in the moose slack channel.
> Do libMesh or Moose have anything that determines the extremes of a variable with out iterating thought the mesh or the solution value?

Dr. Stogner got back to me:

> `libMesh::System::calculate_norm()` with an `L_INF` norm would get you the value of the extreme of the absolute value of a variable (or a weighted combination of variables, depending on how complex the norm you pass in is).  That's the closest thing I can think of, though.  (and of course under the hood it has to iterate through the mesh)

gonuke commented on tutorials/clustering/value_range/example.i:
```diff
@@ -0,0 +1,94 @@
+[]
+
+[UserObjects]
+  [threhsold]
+    type = ThresholdHeuristicsUserObject
+    metric_variable_name = 'metric_var'
+    threshold = 1.1
+  []
+  [value_diff]
+    type = ValueDifferenceHeuristicUserObject
+    metric_variable_name = 'metric_var'
+    tolerance = 0.001
+  []
```
Do we need all of these, or can we do something simpler?

gonuke reviewed (changes requested):
A few final comments

magnoxemo committed:
minor bug fix

magnoxemo committed:
Merge branch 'clustering' into other_heuristics

magnoxemo committed:
minor bug fix

magnoxemo committed:
addressing suggestion

magnoxemo commented:
Fixed some bugs from the previous PR. I’ve added the files here for review as well. Turns out `constexpr` complains because string is not a literal type (which I only found out now).

magnoxemo committed:
another typo

gonuke commented on include/userobjects/BooleanComboClusteringUserObject.h:
```diff
@@ -53,9 +53,9 @@ class BooleanComboClusteringUserObject : public GeneralUserObject
 
   /// operator predecence
-  static const std::unordered_map<std::string, int> _precedence { {"not", 3}, {"!", 3}, {"and", 2}, {"&&", 2}, {"or", 1}, {"||", 1} };
+  std::unordered_map<std::string, int> _precedence { {"not", 3}, {"!", 3}, {"and", 2}, {"&&", 2}, {"or", 1}, {"||", 1} };
```
I think this should still be static, but we may need to define at the top of the implementation (.C) file

gonuke replied:
I think this should still be static, but we may need to define at the top of the implementation (.C) file

gonuke commented on include/userobjects/BooleanComboClusteringUserObject.h:
```diff
@@ -53,9 +53,9 @@ class BooleanComboClusteringUserObject : public GeneralUserObject
   static constexpr int NOT_VISITED = -1;
-  static constexpr std::string _left_parenthesis = "(";
-  static constexpr std::string _right_parenthesis = ")";
+  const  std::string _left_parenthesis = "(";
```
as above - these should still be static, but may need to be defined elsewhere.

gonuke commented on include/userobjects/ValueRangeHeuristicUserObject.h:
```diff
@@ -0,0 +1,24 @@
+
+protected:
+
+  const Real _tolerance_percentage;
+  Real _value;
+  const Real _upper_limit;
+  const Real _lower_limit;
```
How can these be `const` if they are initialized in the constructor?

gonuke commented on src/userobjects/ValueFractionHeuristicUserObject.C:
```diff
@@ -0,0 +1,53 @@
+  : ClusteringHeuristicUserObjectBase(params),
+    _upper_fraction(getParam<Real>("upper_fraction")),
+    _lower_fraction(getParam<Real>("lower_fraction")),
+    _max(std::numeric_limits<Real>::max()),
+    _min(std::numeric_limits<Real>::min())
```
Aren't these initialized backwards?  How will anything ever be larger than `max` to update it?

gonuke commented on src/userobjects/ValueFractionHeuristicUserObject.C:
```diff
@@ -0,0 +1,53 @@
+  for (auto & elem : _mesh.active_element_ptr_range())
+  {
+    score = getMetricData(elem);
+    if (_max < score)
+    {
+      _max = score;
+    }
+    if (_min > score)
+    {
+      _min = score;
+    }
```
Can't we have some version of... ? :
```suggestion
    _max = std::max(_max, score);
    _min = std::min(_min, score);
```

gonuke reviewed (changes requested):
I still think we want some values to be `static` even if they can't be `const`.  Also a few other simplifications.

magnoxemo committed:
fix bugs

magnoxemo committed:
fix input files

gonuke commented on src/userobjects/BooleanComboClusteringUserObject.C:
```diff
@@ -38,7 +46,7 @@ BooleanComboClusteringUserObject::initializeUserObjects()
   for (const auto & token : _output_stack)
   {
-    if (_precedence.find(token))
+    if (_precedence.count(token))
```
This makes a lot more sense - good fix!

gonuke reviewed (approved):
This looks good @magnoxemo - thanks for the great work!

gonuke commented:
I think is is all well done @magnoxemo - I'll see if @aprilnovak or @nuclearkevin have any comments before merging

nuclearkevin commented on doc/content/source/userobjects/ValueDifferenceHeuristicUserObject.md:
```diff
@@ -0,0 +1,21 @@
+elements are less than a tolerance limit.
+
+!alert! warning
+`ValueDifferenceHeuristicUserObejct` doesn't implement the clustering and mesh walking process.
```
I'd recommend mentioning the class(es) that implement the clustering / walking process, just so people know where to find them.

nuclearkevin replied:
I'd recommend pointing the user to user object base which implements the clustering / mesh walking process. This way users know what their options are for combining heuristics with clustering user objects.

This comment applies to all heuristics you added here.

nuclearkevin commented on doc/content/source/userobjects/ValueFractionHeuristicUserObject.md:
```diff
@@ -0,0 +1,28 @@
+$S_{min}$ are maximum and minimum value of the `metric_variable`
+
+!alert! warning
+`ValueFractionHeuristicUserObejct` doesn't implement the clustering and mesh walking process.
```
See comment above.

nuclearkevin commented on doc/content/source/userobjects/ValueFractionHeuristicUserObject.md:
```diff
@@ -0,0 +1,28 @@
+!alert-end!
+
+The following input block declares a `ValueFractionHeuristicUserObejct` which
+is then user by [BooleanComboClusteringUserObject](BooleanComboClusteringUserObject.md)
```
```suggestion
is then used by [BooleanComboClusteringUserObject](BooleanComboClusteringUserObject.md)
```

nuclearkevin commented on include/userobjects/BooleanComboClusteringUserObject.h:
```diff
@@ -53,9 +53,9 @@ class BooleanComboClusteringUserObject : public GeneralUserObject
   static constexpr int NOT_VISITED = -1;
-  static constexpr std::string _left_parenthesis = "(";
-  static constexpr std::string _right_parenthesis = ")";
+  static const std::string _left_parenthesis;
```
Is there any reason why these are `const` instead of `constexpr`?
Same goes for the constants below and above.

magnoxemo replied:
Not sure constexpr works with non literal type.  Compiler complains in this case.

```
BooleanComboClusteringUserObject.h:60:32: error: the type 'const std::string' {aka 'const std::__cxx11::basic_string<char>'} of 'constexpr' variable 'BooleanComboClusteringUserObject::_right_parenthesis' is not literal
   60 |   static constexpr std::string _right_parenthesis = ")";
      |                                ^~~~~~~~~~~~~~~~~~
```

nuclearkevin replied:
Ah, right (forgot about that requirement). You could use a `std::string_view` instead of a `std::string`

nuclearkevin commented on doc/content/source/userobjects/ValueRangeHeuristicUserObject.md:
```diff
@@ -0,0 +1,27 @@
+
+
+!alert! warning
+`ValueRangeHeuristicUserObejct` doesn't implement the clustering and mesh walking process.
```
See previous comment.

nuclearkevin commented on doc/content/source/userobjects/ValueFractionHeuristicUserObject.md:
```diff
@@ -0,0 +1,28 @@
+`ValueFractionHeuristicUserObejct` inherits from the [ClusteringHeuristicUserObjectBase](ClusteringHeuristicUserObjectBase.md).
+It implements the clustering heuristic if the scores of two adjacent
+elements are both either more/less than a upper/lower cut off of the metric variable.
+The upper and lower cut off is calculated is calculated as
```
```suggestion
The upper and lower cut off are calculated as
```

nuclearkevin commented on include/userobjects/BooleanComboClusteringUserObject.h:
```diff
@@ -53,9 +53,9 @@ class BooleanComboClusteringUserObject : public GeneralUserObject
   static constexpr int NOT_VISITED = -1;
-  static constexpr std::string _left_parenthesis = "(";
-  static constexpr std::string _right_parenthesis = ")";
+  static const std::string _left_parenthesis;
```
Why are `_left_parenthesis` and `_right_parenthesis` `static const` instead of `static_constexpr`?

nuclearkevin commented on include/userobjects/ValueDifferenceHeuristicUserObject.h:
```diff
@@ -0,0 +1,18 @@
+
+#include "ClusteringHeuristicUserObjectBase.h"
+
+class ValueDifferenceHeuristicUserObject : public ClusteringHeuristicUserObjectBase
```
Add doxygen comment for class documentation for all of the classes you're adding.

nuclearkevin commented on include/userobjects/ClusteringHeuristicUserObjectBase.h:
```diff
@@ -16,16 +16,16 @@ class ClusteringHeuristicUserObjectBase : public GeneralUserObject
    * A purely virtual function which must be overrided in derived classes.
    * It applies the clustering logic for two elements in the derived class
    */
   virtual bool evaluate(libMesh::Elem * base_element, libMesh::Elem * neighbor_elem) const = 0;
```
Add doxygen comments for the input variables and return type for all of the functions you're adding in this PR. As an example for this function (edit this if it's not accurate):
```c++
/**
   * A purely virtual function which must be overrided in derived classes.
   * It applies the clustering logic for two elements in the derived class
   * param[in] base_element the current element
   * param[in] neighbor_elem the current neighbour of base_element
   * return whether the two elements should be added to a cluster or not
   */
```
For more examples you can check Cardinal's codebase.

nuclearkevin commented on src/userobjects/ValueFractionHeuristicUserObject.C:
```diff
@@ -0,0 +1,48 @@
+
+  Real base_score = getMetricData(base_element);
+  Real neighbor_score = getMetricData(neighbor_element);
+  return (base_score > _upper_cut_off and neighbor_score > _upper_cut_off) ||
```
MOOSE codebases use `&&` instead of the written `and` keyword.

```suggestion
  return (base_score > _upper_cut_off && neighbor_score > _upper_cut_off) ||
```

nuclearkevin commented on src/userobjects/ValueFractionHeuristicUserObject.C:
```diff
@@ -0,0 +1,48 @@
+  Real base_score = getMetricData(base_element);
+  Real neighbor_score = getMetricData(neighbor_element);
+  return (base_score > _upper_cut_off and neighbor_score > _upper_cut_off) ||
+         (base_score < _lower_cut_off and neighbor_score < _lower_cut_off);
```
```suggestion
         (base_score < _lower_cut_off && neighbor_score < _lower_cut_off);
```

nuclearkevin commented on src/userobjects/ValueRangeHeuristicUserObject.C:
```diff
@@ -0,0 +1,42 @@
+ValueRangeHeuristicUserObject::isInsideTheRange(libMesh::Elem * element) const
+{
+  Real score = getMetricData(element);
+  return _lower_limit < score and score < _upper_limit;
```
```suggestion
  return _lower_limit < score && score < _upper_limit;
```

nuclearkevin commented on src/userobjects/ValueRangeHeuristicUserObject.C:
```diff
@@ -0,0 +1,42 @@
+ValueRangeHeuristicUserObject::evaluate(libMesh::Elem * base_element,
+                                        libMesh::Elem * neighbor_element) const
+{
+  return isInsideTheRange(base_element) and isInsideTheRange(neighbor_element);
```
```suggestion
  return isInsideTheRange(base_element) && isInsideTheRange(neighbor_element);
```

nuclearkevin commented on tutorials/clustering/value_difference/example.i:
```diff
@@ -0,0 +1,77 @@
+# mesh_inputs_params
```
All of these parameters are used once. I'd recommend placing them in the `[Mesh]` block instead of using static input file variables to shrink the length of this test (and the other tests). 

The longer the input file, the more confusing it can be for new uses (especially if they need to hop around to figure out where variables are coming from).

nuclearkevin reviewed (changes requested):
Sorry for the wait on this! I don't have any substantial questions / comments on the heuristics themselves (my comments are mostly style related). 

I'll get to reviewing your online clustering PRs over the next few days, but I imagine some of the comments I've left here might apply to those as well.

magnoxemo committed:
addressing suggestions from Kevin

magnoxemo commented:
As per my online discussion with April, I am merging it here and then I will open a PR with tests and gold files to the `cardinal/devel` branch.
